apps/meteor/server/startup/migrations/v320.ts

```diff
@@ -0,0 +1,93 @@
+import type { IRole, IUser } from '@rocket.chat/core-typings';
+import { ROOM_ROLE_PRIORITY_MAP } from '@rocket.chat/core-typings';
+import { Users, Subscriptions } from '@rocket.chat/models';
+
+import { addMigration } from '../../lib/migrations';
+
+export const getRoomRolePriorityForRole = (role: IRole['_id']) =>
+	(ROOM_ROLE_PRIORITY_MAP as { [key: IRole['_id']]: number })[role] ?? ROOM_ROLE_PRIORITY_MAP.default;
+
+export const calculateRoomRolePriorityFromRoles = (roles: IRole['_id'][]) => {
+	return roles.reduce((acc, roleId) => Math.min(acc, getRoomRolePriorityForRole(roleId)), ROOM_ROLE_PRIORITY_MAP.default);
+};
+
+const READ_BATCH_SIZE = 1000;
+const WRITE_BATCH_SIZE = 1000;
+
+async function assignRoomRolePrioritiesFromMap(userIdAndroomRolePrioritiesMap: Map<IUser['_id'], IUser['roomRolePriorities']>) {
+	let bulk = Users.col.initializeUnorderedBulkOp();
+	let counter = 0;
+
+	for await (const [userId, roomRolePriorities] of userIdAndroomRolePrioritiesMap.entries()) {
+		userIdAndroomRolePrioritiesMap.delete(userId);
+
+		if (roomRolePriorities) {
+			const updateFields = Object.entries(roomRolePriorities).reduce(
+				(operations, rolePriorityData) => {
+					const [rid, rolePriority] = rolePriorityData;
+					operations[`roomRolePriorities.${rid}`] = rolePriority;
+					return operations;
+				},
+				{} as Record<string, number>,
+			);
+
+			bulk.find({ _id: userId }).updateOne({
+				$set: updateFields,
+			});
+
+			counter++;
+
+			if (counter >= WRITE_BATCH_SIZE) {
+				// Execute the bulk operation for the current batch
+				await bulk.execute();
+				bulk = Users.col.initializeUnorderedBulkOp();
+				counter = 0;
+			}
+		}
+	}
+
+	// Execute any remaining operations
+	if (counter > 0) {
+		await bulk.execute();
+	}
+
+	return userIdAndroomRolePrioritiesMap;
+}
+
+addMigration({
+	version: 320,
+	name: 'Add "roomRolePriorities" field to users',
+	async up() {
+		let userIdAndroomRolePrioritiesMap = new Map<IUser['_id'], IUser['roomRolePriorities']>();
+
+		// Create a cursor with the specified batch size
+		const cursor = Subscriptions.find(
+			{ roles: { $exists: true } },
+			{
+				projection: { 'rid': 1, 'roles': 1, 'u._id': 1 },
+				sort: { _id: 1 },
+			},
+		).batchSize(READ_BATCH_SIZE);
+
+		for await (const sub of cursor) {
+			if (!sub.roles?.length) {
+				continue;
+			}
+
+			const userId = sub.u._id;
+			const roomId = sub.rid;
+			const priority = calculateRoomRolePriorityFromRoles(sub.roles);
+
+			const existingPriorities = userIdAndroomRolePrioritiesMap.get(userId) || {};
+			existingPriorities[roomId] = priority;
+			userIdAndroomRolePrioritiesMap.set(userId, existingPriorities);
+
+			if (userIdAndroomRolePrioritiesMap.size >= WRITE_BATCH_SIZE) {
+				userIdAndroomRolePrioritiesMap = await assignRoomRolePrioritiesFromMap(userIdAndroomRolePrioritiesMap);
+			}
+		}
+
+		// Flush any remaining priorities in the map
+		await assignRoomRolePrioritiesFromMap(userIdAndroomRolePrioritiesMap);
```

Comment:
Do you still need this? I thought you could initialize an unorderedop and just push things on there, and then consume the operation (since mongo says it groups it efficiently) so you don't have to do the manual batching 👀